src/range.cr

```diff
@@ -252,6 +252,54 @@ struct Range(B, E)
     {% end %}
   end
 
+  # Returns the union of this range, and another.
+  # Returns `self.begin...self.begin` if there is no overlap.
+  #
+  # ```
+  # (1..5).union(5..10)   # => 1..10
+  # (1...10).union(1..10) # => 1..10
+  # ```
+  def union(other : Range)
+    e_s : E = self.end
+    e_o : E = other.end
+
+    return Range.new(self.begin, self.begin, exclusive: true) if self.end < other.begin && (!e_s.responds_to?(:succ) || e_s.succ != other.begin)
+    return Range.new(self.begin, self.begin, exclusive: true) if other.end < self.begin && (!e_o.responds_to?(:succ) || e_o.succ != self.begin)
+
+    Range.new(
+      Math.min(self.begin, other.begin),
+      Math.max(self.end, other.end),
+      {self, other}.max_by(&.end).excludes_end?
+    )
+  end
+
+  # Returns the intersection of this range, and another.
+  # Returns `self.begin...self.begin` if there is no overlap.
+  #
+  # ```
+  # (2..10).intersection(0..8)   # => 2..8
+  # (1...10).intersection(7..12) # => 7...10
+  # ```
+  def intersection(other : Range)
+    return Range.new(self.begin, self.begin, exclusive: true) if self.end < other.begin || other.end < self.begin
```

Comment:
An intersection is empty when there is no overlap.

This refactor is actually how I uncovered the bug in `overlaps?` for the very slightly overlapping integer ranges. I totally would've missed it otherwise.

```suggestion
    return Range.new(self.begin, self.begin, exclusive: true) unless overlaps?(other)
```